### PR TITLE
take into account optional cron within datasets

### PR DESCRIPTION
**Pull request by fupelaqu**
## Summary
The goal of this PR is to support data lineage freshness in dags generated by Starlake for Airflow.

Data lineage is implemented within Airflow by using data-aware scheduling.

Each time a load or transformation is executed within a dag, Starlake will produce the corresponding Dataset using outlets.

Similarly, when a transformation requires that other load(s) or transformation(s) be performed beforehand, Starlake will add them as a list of Datasets to the corresponding consumer dag so that the dag will be automatically triggered each time all the datasets would have been produced .

The problems start when a dag has dependencies on Datasets that are produced by dag(s) whose frequency(ies) may differ. 

For example a dag may depend on one Dataset A that is produced once a month and another Dataset B that is  produced daily... 
At the end the dag will be triggered not every day as expected but not before the dag that produces Dataset A will be triggered, which means once a month.

The freshness of the data will ultimately be defined by the dataset updated the least frequently.

The solution consists of ensuring that all datasets produced by a dag for which a cron has been defined are republished identically by the consumer dag.

This solution implies : 

- adding the optional schedule (cron) for each dependency computed for dags
- adding to the uri of each dataset produced by a dag for which a cron has been defined, in the form of a query parameter (sl_schedule), the logical date time of the dag
- adding to the uri of each dataset consumed by a dag for which the corresponding dependency has defined a cron, in the form of a query parameter (sl_schedule), the last logical date time the dataset should have been produced
- republishing those scheduled datasets

**PR Type: Feature**

**Status: Ready to review**

**Breaking change? No**

### CODE REVIEW

**Review by mhdkassir — APPROVED** [review]
Lgtm

File: src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_cloud_run_job.py [hunk 1]
```diff
@@ -35,6 +36,7 @@ def __init__(
         self.cloud_run_job_name = __class__.get_context_var(var_name='cloud_run_job_name', options=self.options) if not cloud_run_job_name else cloud_run_job_name
         self.cloud_run_job_region = __class__.get_context_var('cloud_run_job_region', "europe-west1", self.options) if not cloud_run_job_region else cloud_run_job_region
         self.cloud_run_async = __class__.get_context_var(var_name='cloud_run_async', default_value="True", options=self.options).lower() == "true" if cloud_run_async is None else cloud_run_async
+        self.cloud_run_async_poke_interval = float(__class__.get_context_var('cloud_run_async_poke_interval', "30", self.options)) if not cloud_run_async_poke_interval else cloud_run_async_poke_interval
```
Comment (zedach) [inline src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_cloud_run_job.py:39]: this changes seems to be related to cloud run jobs feature, I'm I wrong ?

File: src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_cloud_run_job.py [hunk 2]
```diff
@@ -50,7 +52,7 @@ def __job_with_completion_sensors__(self, task_id: str, command: str, spark_conf
                     f"gcloud beta run jobs execute {self.cloud_run_job_name} "
                     f"--args \"{command}\" "
                     f"{self.update_env_vars} "
-                    f"--async --region {self.cloud_run_job_region} --project {self.project_id} --format='get(metadata.name)'" #--task-timeout 300 
+                    f"--async --region {self.cloud_run_job_region} --project {self.project_id} --format='get(metadata.name)'" #--task-timeout 300
```
Comment (zedach) [inline src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_cloud_run_job.py:55]: also Here. is this related to optional cron within dataset ?

File: src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_dataproc_job.py [hunk 3]
```diff
@@ -64,29 +64,31 @@ def __init__(self, cluster_config: StarlakeAirflowDataprocClusterConfig, options
...
+            task_id: str=None,
+            cluster_name: str=None,
+            **kwargs) -> BaseOperator:
         """
         Create the Cloud Dataproc cluster.
         This operator will be flagged a success if the cluster by this name already exists.
         """
         cluster_id = self.cluster_config.cluster_id if not cluster_id else cluster_id
-        cluster_name = f"{self.cluster_config.dataproc_name}-{cluster_id.replace('_', '-')}-{TODAY}" if not cluster_name else cluster_name
+        cluster_name = f"{self.cluster_config.dataproc_name}-{cluster_id.replace('_', '-')}-{TODAY}"[0:51] if not cluster_name else cluster_name[0:51]
```
Comment (zedach) [inline src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_dataproc_job.py:89]: Same here in this file: change seems not to be related to the feature but with dataproc cluster name limitation. maybe should be in a fix PR, what do you think ?

File: src/main/resources/templates/dags/__common_airflow.py.j2 [hunk 4]
```diff
@@ -22,3 +22,23 @@ def default_spark_config(*args, **kwargs) -> StarlakeSparkConfig:
...
+    if parameters.__len__() > 0 :
+        return '?' + '&'.join(list(f'{parse.quote(k)}={parse.quote(v)}' for (k,v) in parameters.items()))
+    else :
+        return ''
+
+from croniter import croniter
+from datetime import datetime as dt
+
+def cron_start_time() -> dt:
+    return dt.fromtimestamp(dt.now().timestamp())
```
Comment (zedach) [inline src/main/resources/templates/dags/__common_airflow.py.j2:38]: Why not using `dt.now()`  directly to obtain the datetime object ?

File: src/main/resources/templates/dags/__common_airflow.py.j2 [hunk 5]
```diff
@@ -22,3 +22,23 @@ def default_spark_config(*args, **kwargs) -> StarlakeSparkConfig:
         **kwargs
     )
 spark_config = getattr(sys.modules[__name__], "get_spark_config", default_spark_config)
+
+from urllib import parse
```
Comment (zedach) [inline src/main/resources/templates/dags/__common_airflow.py.j2:26]: maybe import quote directly simplify the code 
`from urllib.parse import quote`

**Review by zedach — CHANGES_REQUESTED** [review]
LGTM. Some changes are due to formatting and other corrections that don't seem to be related to the new functionality. I propose to include them in another PR if that's the case

**Review by zedach — APPROVED** [review]

File: src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_cloud_run_job.py (hunk 1 above)
Comment (fupelaqu) [reply src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_cloud_run_job.py:39]: it is right but it has been added in this PR while testing the latter

File: src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_dataproc_job.py (hunk 3 above)
Comment (fupelaqu) [reply src/main/python/starlake-airflow/ai/starlake/airflow/gcp/starlake_airflow_dataproc_job.py:89]: You are totally right this update is related with dataproc cluster nae limitation and has been added in this PR while testing the latter. I don't think we should create a specific PR for this

**Pull request merged**